File: iteration-2/src/Student.java
import java.util.ArrayList;

public class Student extends Person {
    private String studentID;
    private Semester semester;
    private Advisor advisor;
    private Transcript transcript;
    private ArrayList<Course> selectedCourses;
    private CourseSession selectedSessions;
    private int entryYear;
    private Schedule schedule;
    private float gpa;
    private float cgpa;
    private int completedCredit;
    private List<Course> activeCourses;
    private List<Course> pastCourses;
    private List<Course> nonTakenCourses;
    private List<Course> failedCourses;

    public Student(String studentID, Semester semester, Advisor advisor, Transcript transcript, ArrayList<Course> selectedCourses,
                   CourseSession selectedSessions, int entryYear, Schedule schedule, float gpa, float cgpa, int completedCredit,
                   List<Course> activeCourses, List<Course> pastCourses, List<Course> nonTakenCourses, List<Course> failedCourses,) {
        this.studentID = studentID;
        this.semester = semester;
        this.advisor = advisor;
        this.transcript = transcript;
        this.selectedCourses = selectedCourses;
        this.selectedSessions = selectedSessions;
        this.entryYear = entryYear;
        this.schedule = schedule;
        this.gpa = gpa;
        this.cgpa = cgpa;
        this.completedCredit = completedCredit;
        this.activeCourses = activeCourses;
        this.pastCourses = pastCourses;
        this.nonTakenCourses = nonTakenCourses;
        this.failedCourses = failedCourses;
    }


    // getter and setter methods
    public String getStudentID() {return this.studentID;}
    public void setStudentID(String studentID) {this.studentID = studentID;}

    public Semester getSemester() {return this.semester;}

    public void setSemester(Semester semester) {this.semester = semester;}

    public Advisor getAdvisor() {return this.advisor;}

    public void setAdvisor(Advisor advisor) {this.advisor = advisor;}

    public Transcript getTranscript() {return this.transcript;}

    public void setTranscript(Transcript transcript) {this.transcript = transcript;}

    public ArrayList<Course> getSelectedCourses() {return this.selectedCourses;}

    public void setSelectedCourses(ArrayList<Course> selectedCourses) {this.selectedCourses = selectedCourses;}

    public CourseSession getSelectedSessions() {return this.selectedSessions;}

    public void setSelectedSessions(CourseSession selectedSessions) {this.selectedSessions = selectedSessions;}

    public int getEntryYear() {return this.entryYear;}

    public void setEntryYear(int entryYear) {this.entryYear = entryYear;}

    public Schedule getSchedule() {return this.schedule;}

    public void setSchedule(Schedule schedule) {this.schedule = schedule;}

    public float getGpa() {return gpa;}

    public void setGpa(float gpa) {this.gpa = gpa;}

    public float getCgpa() {return cgpa;}

    public void setCgpa(float cgpa) {this.cgpa = cgpa;}

    public int getCompletedCredit() {return completedCredit;}

    public void setCompletedCredit(int completedCredit) {this.completedCredit = completedCredit;}

    public List<Course> getActiveCourses() {return activeCourses;}

    public void setActiveCourses(List<Course> activeCourses) {this.activeCourses = activeCourses;}

    public List<Course> getPastCourses() {return pastCourses;}

    public void setPastCourses(List<Course> pastCourses) {this.pastCourses = pastCourses;}

    public List<Course> getNonTakenCourses() {return nonTakenCourses;}

    public void setNonTakenCourses(List<Course> nonTakenCourses) {this.nonTakenCourses = nonTakenCourses;}

    public List<Course> getFailedCourses() {return failedCourses;}

    public void setFailedCourses(List<Course> failedCourses) {this.failedCourses = failedCourses;}

    public void addCourseToSelectedCourses(Course course) {
        this.selectedCourses.add(course);
    }

    public void calculateGpa(){
        int credit = transcript.getCredit();
        int grade = transcript.getGrade();
        int result = grade/credit;
        setGpa((float) result);
    }

    public void calculateCumulativeGpa(){
        int cCredit = transcript.getCumulativeCredit();
        int cGrade = transcript.getCumulativeGrade();
        int result = cGrade/cCredit;
        setCgpa((float) result);
    }

    public void printActiveCourses(){
        int i = 0;
        System.out.println(getName()+' '+getSurname());
        while (i < activeCourses.size()) {
            Course course = activeCourses.get(i);
            System.out.println(course.getName());
            i++;
        }
    }
    @Override
    public String toString() {
        return "Student{" +
                "advisor=" + advisor +
                "} " + super.toString();
    }

}

File: iteration-3/Student.py
from typing import List
from Person import Person
from Semester import Semester
from Transcript import Transcript
from Course import *
from CourseSession import CourseSession
from CollisionChecker import *

class Student(Person):
    def __init__(self, name: str, surname: str, id: int, semester: Semester, transcript: Transcript, selectedCourses: List[Course], selectedSessions: CourseSession, entryYear: int, schedule: List[list], activeCourses: List[Course],  nonTakenCourses: List[Course]):
        super().__init__(name, surname, id)
        self._semester = semester
        self._transcript = transcript
        self._selectedCourses = selectedCourses
        self._selectedSessions = selectedSessions
        self._entryYear = entryYear
        self._schedule = schedule
        self._activeCourses = activeCourses
        self._nonTakenCourses = nonTakenCourses
        self._password = "123"

    def get_full_name(self):
        return f"{self._name} {self._surname}"

    
    # Getter and setter methods for the semester attribute
    @property
    def semester(self):
        return self._semester

    @semester.setter
    def semester(self, value: Semester):
        self._semester = value

    # Getter and setter methods for the transcript attribute
    @property
    def transcript(self):
        return self._transcript

    @transcript.setter
    def transcript(self, value: Transcript):
        self._transcript = value
    
    # Getter and setter methods for the selectedCourses attribute
    @property
    def selectedCourses(self):
        return self._selectedCourses

    @selectedCourses.setter
    def selectedCourses(self, value: List[Course]):
        self._selectedCourses = value

    # Getter and setter methods for the selectedSessions attribute
    @property
    def selectedSessions(self):
        return self._selectedSessions

    @selectedSessions.setter
    def selectedSessions(self, value: CourseSession):
        self._selectedSessions = value

    # Getter and setter methods for the entryYear attribute
    @property
    def entryYear(self):
        return self._entryYear

    @entryYear.setter
    def entryYear(self, value: int):
        self._entryYear = value
     

    # Getter and setter methods for the activeCourses attribute
    @property
    def activeCourses(self):
        return self._activeCourses
    
    @activeCourses.setter
    def activeCourses(self, value: List[Course]):
        self._activeCourses = value

    # Getter and setter methods for the nonTakenCourses attribute
    @property
    def nonTakenCourses(self):
        return self._nonTakenCourses

    @nonTakenCourses.setter
    def nonTakenCourses(self, value: List[Course]):
        self._nonTakenCourses = value

    # Getter and setter methods for the failed_courses attribute
    @property
    def failedCourses(self):
        return self._failedCourses

    @failedCourses.setter
    def failedCourses(self, value: List[Course]):
        self._failedCourses = value

    

def calculateCumulativeGpa(self):
    try:
        transcript = Transcript()
        cCredit = transcript.cumulativeCredit
        cGrade = transcript.cumulativeCredit
        result = cGrade/cCredit
        self._cgpa = result
    except Exception as e:
        print(e)
    except ZeroDivisionError as z:
        print(z)
    
    def printNonTakenCourses(self):
        try:
            for course in self._nonTakenCourses:
                print(course)
        except Exception as e:
            print(e)

    def printFailedCourses(self):
        try:
            for course in self._failedCourses:
                print(course)
        except Exception as e:
            print(e)

    def printSelectedSessions(self):
        try:
            for courseSession in self._selectedSessions:
                print(courseSession)
        except Exception as e:
            print(e)


    def addNonTakenCourse(self,course: Course):
        try:
            if course not in self.nonTakenCourses:
                self.nonTakenCourses.append(course)
        except Exception as e:
            print(e)


    def __str__(self):
        return f"Student(name='{self._name}', surname='{self._surname}', id={self._id}, emails='{self._email}', semester='{self._semester}',  transcript='{self._transcript}', selected_courses='{self._selectedCourses}', selected_sessions='{self._selectedSessions}', entry_year={self._entryYear}, schedule='{self._schedule}', cgpa={self._cgpa}, active_courses='{self._activeCourses}', past_courses='{self._pastCourses}', non_taken_courses='{self._nonTakenCourses}', failed_courses='{self._failedCourses}')"
